### ISSUE TITLE
Array syntax idea

### ISSUE THREAD

**Issue by MarcusJohnson91** [toml-lang/toml#1031]
Name = PNGTests
NumCases = 1

[Case.0]
State = TestState_Enabled
Outcome = Outcome_Passed
Size = 138 # Num hexadecimal digits, divide by 2 to get the number of bytes.
Data = 0x[89504E470D0A1A0A0000000D4948445200000001000000010001000000376EF9240000000A49444154000078016360000000020001737501180000000049454E44AE426082]

No whitespace, commas, or anything except hexadecimal digits are allowed after 0x[ and before the closing ].

It makes parsing binary data stored in a toml file MUCH easier and faster to process.

I don’t see why binary and octal couldn’t have the same syntax. 0b[ and 0o[ respectively, but I have no need for such syntax.

**Comment by arp242** [comment]
I assume what you want is that this:

    data = 0x[89 50 4e]

will be treated like:

    data = "\x89\x50\x4e"

Or something along these lines.

If I had a lot of binary data, I'd just put in in a string like:

    data = '''
    89 50 4E 47 0D 0A 1A 0A 00 00 00 0D
    49 48 44 52 …
    '''

And then write a custom parser in your language to deal with that. Should be easy enough in most languages.

Or use an array of numbers, or escapes like above.

This seems far too rare of a use case to add to TOML.

At the very least we'd need a few examples of real-world TOML files that see actual use where this feature would be useful.

**Comment by eksortso** [comment]
Remember that parsers treat things like `"\x89\x50\x4e"` not as byte arrays. They're strings. In fact, `"\x89" on its own is two bytes when encoded as UTF-8.

The custom parser idea makes more sense in context, but to be honest, I need more context. We do see more computer-generated values, like hashes, in use cases where TOML is not touched by humans. But invariably that binary data is expressed as a hex string. It's similar to the proposed `0x[]` syntax in that way. I'm not opposed to a byte array value type, but I'm skeptical that we need it everywhere.

Send us more use cases where byte arrays need their own special syntax. Why would a human-centric format need a value type that emits blobs of arbitrary binary data and skips the checks available to an intermediate string format?

**Comment by MarcusJohnson91** [comment]
Where are you guys getting the idea that it’s a string?

granted, technically the entire TOML file is a string.

but beyond that it’s just back to back hex digits after the 0x[ and before the ].

No spaces, no escapes. Just hexadecimal digits.

**Comment by arp242** [comment]
> Remember that parsers treat things like "\x89\x50\x4e" not as byte arrays. They're strings. In fact, `"\x89" on its own is two bytes when encoded as UTF-8.

Oh yeah, of course 🤦 I've been dealing with a lot of these kind of (non-UTF8) strings this week and in context switching I just forgot TOML doesn't work like that.

**Comment by eksortso** [comment]
> Where are you guys getting the idea that it’s a string?

I did know that what you are talking about are byte arrays. My point was that in TOML, string values are quoted, so custom parsing would still be required to turn such strings into proper byte arrays. This was in response to @arp242, who already replied.

> No spaces, no escapes. Just hexadecimal digits.

We need more use cases for binary arrays before we could proceed with creating new syntax, and I implore you and others to provide those use cases.

But even if we proceeded, I would not sign off on such a limited expression set. I would want to allow ignorable newlines and whitespace between the brackets as well, since human beings might want to use the syntax, and most likely such users would want to avoid long lines and unbroken sequences of hex digits, cmiiw.

**Comment by arp242** [comment]
Based on the above discussion and the lack of real-world examples that would benefit from this, I'm going to go ahead and close this as "not currently planned".

While this might be a bit useful on some occasion, by and large this seems like a rare use case for TOML files, and even for those uses it's not that hard to make it work via an array of numbers, base64, or some other binary-to-text encoding.

The way forward would be to get some real-world examples that me and eksortso asked for. That is: "TOML file A from application B needs to do C for binary data, which is very awkward/difficult without this new syntax". There should ideally be several examples from software that sees some real-world usage. Personally I suspect this doesn't exist, but if they do I'd be willing to change my mind.

**Comment by epage** [comment]
One crazy idea for byte arrays is using hexadecimal bigints

**Comment by ChristianSi** [comment]
@epage But TOML has no bigint type. Even if it had, there would be no advantage over Base-64. If you really feel like it, you can of course write hex numbers of arbitrary size as multiline strings.

**Comment by epage** [comment]
> @epage But TOML has no bigint type.

Maybe I'm misusing a term but TOML leaves the maximum integer size up to the implementation, allowing for implementations to support arbitrary sized integers.

**Comment by ChristianSi** [comment]
@epage: Yes, some implementations may have them, but others won't, so files using integers outside the 64-bit signed range (recommended by the spec as range any implementation *should* support) aren't portable and will likely be rejected with an error by some spec-confirming implementations.
